File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/arch_units/go/au_135_c_eq_u64_f64.go
// arch-unit 135  --  c  `a == b`  lhs=uint64_t rhs=double
// symbol op_478   outcome LIFTED
//
// the arch-unit, arch-opcode by arch-opcode:
//   fcvt.d.lu fa5, a0, dyn             float    emulation:ui64_to_f64_rm0
//   feq.d a0, fa0, fa5                 float    emulation:f64_eq_rm0
//   c.jr ra                            integer  return
//
// answer: int, 64 bits.  parameters are operand bit patterns.
package archunits

import emul "softfloat_emul"

func Au_135_c_eq_u64_f64(p0 uint64, p1 uint64) uint64 {
	// a0: operand `a` (uint64_t) arrives in a0
	var v1 uint64 = p0
	// fa0: operand `b` (double) arrives in fa0
	var v2 uint64 = p1
	var v3 uint64 = emul.Emu_ui64_to_f64_rm0(v1)
	var v4 uint64 = b2u(emul.Emu_f64_eq_rm0(v2, v3))
	return v4
}

File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/arch_units/c/au_135_c_eq_u64_f64.c
/*
 * arch-unit 135  --  c  `a == b`  lhs=uint64_t rhs=double
 * symbol op_478   outcome LIFTED
 *
 * the arch-unit, arch-opcode by arch-opcode:
 *   fcvt.d.lu fa5, a0, dyn             float    emulation:ui64_to_f64_rm0
 *   feq.d a0, fa0, fa5                 float    emulation:f64_eq_rm0
 *   c.jr ra                            integer  return
 *
 * answer: int, 64 bits.  parameters are operand bit patterns.
 */
#include "arch_units.h"

uint64_t au_135_c_eq_u64_f64(uint64_t p0, uint64_t p1)
{
    /* a0: operand `a` (uint64_t) arrives in a0 */
    const uint64_t v1 = p0;
    /* fa0: operand `b` (double) arrives in fa0 */
    const uint64_t v2 = p1;
    const uint64_t v3 = ui64_to_f64_rm0(v1);
    const uint64_t v4 = au_b2u(f64_eq_rm0(v2, v3));
    return v4;
}
